Add Helm/Operator instructions & Cleanup (#13699)
DataDog/documentation: content/en/agent/cluster_agent/endpointschecks.md

## Overview

The cluster check feature provides the ability to Autodiscover and perform checks on load-balanced cluster services (eg. Kubernetes services). The [service based cluster checks][1] can be used to schedule one instance of the desired check with respect to the service and its IP address. In comparison Endpoint Checks extend this mechanism to monitor *each* endpoint managed by the Kubernetes service.

> The [Cluster Agent][2] discovers the Endpoint Check configurations based on Autodiscovery Annotations on the Kubernetes services, then dispatches them to node-based Agents to individually run. The Endpoint Checks are dispatched to Agents that run on the same node as the pod(s) that back the endpoint(s) of the monitored Kubernetes service. This dispatching logic allows the Agent to add the pod and container tags it has already automatically collected for the respective pod(s).

Reviewer: ```suggestion
The [Cluster Agent][2] discovers the Endpoints Check configurations based on Autodiscovery annotations on the Kubernetes services. The Cluster Agent then dispatches these configurations to node-based Agents to individually run. The Endpoints Checks are dispatched to Agents that run on the same node as the pod(s) that back the endpoint(s) of the monitored Kubernetes service. This dispatching logic allows the Agent to add the pod and container tags it has already automatically collected for the respective pod(s).
```

> The Agents connect to the Cluster Agent every 10 seconds and retrieve the check configurations to run. Metrics coming from endpoints checks are submitted with service tags, [kubernetes tags][3], host tags, and the `kube_endpoint_ip` tag based on the evaluated IP Address.

Reviewer: ```suggestion
The Agents connect to the Cluster Agent every ten seconds and retrieve the check configurations to run. Metrics coming from Endpoint Checks are submitted with service tags, [Kubernetes tags][3], host tags, and the `kube_endpoint_ip` tag based on the evaluated IP address.
```

> This feature is supported on Kubernetes for versions 6.12.0+ of the Agent, and versions 1.3.0+ of the Cluster Agent. Starting with version 1.4.0, the Cluster Agent converts every endpoints check of a non-pod-backed endpoint into a regular cluster check. Enable the [cluster check][4] feature alongside endpoints checks to take advantage of this functionality.

Reviewer: ```suggestion
The Endpoints Checks feature is supported on Kubernetes for Agent v6.12.0+ and Cluster Agent v1.3.0+. Starting with v1.4.0, the Cluster Agent converts every Endpoints Check of a non-pod-backed endpoint into a regular Cluster Check. Enable the [Cluster Check][4] feature alongside Endpoints Checks to take advantage of this functionality.
```

> ### Example: Service with Endpoints

Reviewer: ```suggestion
### Example: Service with endpoints
```

> In this example below a Kubernetes deployment for nginx was created with 3 pods.

Reviewer: ```suggestion
In the following example, a Kubernetes deployment for NGINX was created with three pods.
```

```shell
# kubectl get pods --selector app=nginx -o wide
NAME                     READY   STATUS    RESTARTS   AGE   IP           NODE
nginx-66d557f4cf-m4c7t   1/1     Running   0          3d    10.0.0.117   gke-cluster-default-pool-4658d5d4-k2sn
nginx-66d557f4cf-smsxv   1/1     Running   0          3d    10.0.1.209   gke-cluster-default-pool-4658d5d4-p39c
nginx-66d557f4cf-x2wzq   1/1     Running   0          3d    10.0.1.210   gke-cluster-default-pool-4658d5d4-p39c
```

> A service was also created, linking up this service to the pods through these 3 endpoints.

Reviewer: ```suggestion
A service was also created, linking to the pods through these three endpoints.
```

```shell
# kubectl get service nginx -o wide
NAME    TYPE        CLUSTER-IP     EXTERNAL-IP   PORT(S)   AGE   SELECTOR
nginx   ClusterIP   10.3.253.165   <none>        80/TCP    1h    app=nginx
```

```shell
# kubectl get endpoints nginx -o yaml
...
- addresses:
  - ip: 10.0.0.117
    nodeName: gke-cluster-default-pool-4658d5d4-k2sn
    targetRef:
      kind: Pod
      name: nginx-66d557f4cf-m4c7t
      ...
  - ip: 10.0.1.209
    nodeName: gke-cluster-default-pool-4658d5d4-p39c
    targetRef:
      kind: Pod
      name: nginx-66d557f4cf-smsxv
      ...
  - ip: 10.0.1.210
    nodeName: gke-cluster-default-pool-4658d5d4-p39c
    targetRef:
      kind: Pod
      name: nginx-66d557f4cf-x2wzq
      ...
```

> So while a Service based cluster check tests the Service's single IP Address, the Endpoint Checks are scheduled for *each* of these three endpoints associated with this Service. By design too, the Endpoint Checks are dispatched to Agents that run on the same node as the pods that back the endpoints of this `nginx` service. So in this example the Agents running on the nodes `gke-cluster-default-pool-4658d5d4-k2sn` and `gke-cluster-default-pool-4658d5d4-p39c` run the checks against these `nginx` pods.

Reviewer: ```suggestion
While a Service-based Cluster Check tests the service's single IP address, Endpoints Checks are scheduled for *each* of the three endpoints associated with this service. 

By design, Endpoints Checks are dispatched to Agents that run on the same node as the pods that back the endpoints of this `nginx` service. In this example, the Agents running on the nodes `gke-cluster-default-pool-4658d5d4-k2sn` and `gke-cluster-default-pool-4658d5d4-p39c` run the checks against these `nginx` pods.
```

> ## Set up Endpoint Check dispatching

Reviewer: ```suggestion
## Set up Endpoints Check dispatching
```

{{< tabs >}}
{{% tab "Helm" %}}
This is enabled by default in the Helm deployment of the Cluster Agent through the `datadog.clusterChecks.enabled` configuration key:
```yaml
datadog:
  clusterChecks:
    enabled: true
  # (...)
clusterAgent:
  enabled: true
  # (...)
```

> This configuration enables both cluster check and Endpoint Check dispatching, between the Cluster Agent and the Agents.

Reviewer: ```suggestion
This configuration enables Cluster Check and Endpoints Check dispatching between the Cluster Agent and the Agents.
```

{{% /tab %}}
{{% tab "Operator" %}}

> Cluster check dispatching is enabled in the Operator deployment of the Cluster Agent by using the `clusterAgent.config.clusterChecksEnabled` configuration key:

Reviewer: ```suggestion
Cluster Check dispatching is enabled in the Operator deployment of the Cluster Agent by using the `clusterAgent.config.clusterChecksEnabled` configuration key:
```

```yaml
apiVersion: datadoghq.com/v1alpha1
kind: DatadogAgent
metadata:
  name: datadog
spec:
  # (...)
  clusterAgent:
    config:
      clusterChecksEnabled: true
```

> This configuration enables both cluster check and Endpoint Check dispatching, between the Cluster Agent and the Agents.

Reviewer: ```suggestion
This configuration enables Cluster Check and Endpoints Check dispatching between the Cluster Agent and the Agents.
```

DD_EXTRA_LISTENERS="kube_endpoints"
```

> **Note**: If the monitored endpoints are not backed by pods, you must [enable cluster checks][1]. This can be done by adding the `kube_services` configuration provider and listener:

Reviewer: ```suggestion
**Note**: If the monitored endpoints are not backed by pods, you must [enable Cluster Checks][1]. This can be done by adding the `kube_services` configuration provider and listener:
```

          polling: true
    ```

> **Note**: If the monitored endpoints are not backed by pods, you must [enable cluster checks][1]. This can be done by adding the `clusterchecks` configuration provider:

Reviewer: ```suggestion
**Note**: If the monitored endpoints are not backed by pods, you must [enable Cluster Checks][1]. This can be done by adding the `clusterchecks` configuration provider:
```

ad.datadoghq.com/endpoints.logs: '[<LOGS_CONFIG>]'
```

The `%%host%%` [template variable][6] is supported and is replaced by the endpoints' IPs. The `kube_namespace`, `kube_service`, and `kube_endpoint_ip` tags are automatically added to the instances.

**Note:** The custom endpoints logs configuration is only supported during Docker socket log collection and not Kubernetes log file collection.

### Unified Service Tagging
You can optionally add the standard labels to the Kubernetes service for [Unified Service Tagging][7] to set the `env`, `service`, and `version` tags on data generated by these checks.
```yaml
tags.datadoghq.com/env: "<ENV>"
tags.datadoghq.com/service: "<SERVICE>"
tags.datadoghq.com/version: "<VERSION>"
```

#### Example: HTTP check on an NGINX-backed service with NGINX check on the service's endpoints

> The example below takes advantage of all these options. This Service is associated to the pods of the `nginx` deployment. Based on this configuration:

Reviewer: ```suggestion
The example below takes advantage of all these options. This service is associated with the pods of the `nginx` deployment. Based on this configuration:
```

> - An [`nginx`][8] based endpoint check will be dispatched for each Nginx pod backing this Service, this `nginx` check will be ran by the Agents on the same respective nodes as the Nginx pods using the Pod IP as the `%%host%%`.

Reviewer: ```suggestion
- An [`nginx`][8] based Endpoints Check is dispatched for each NGINX pod backing this service. Agents on the same respective nodes as the NGINX pods (using the pod IP as `%%host%%`) run this check. 
```

> - An [`http_check`][9] based cluster check will be dispatched to a single Agent in the cluster, this `http_check` will use the IP of the Service as the `%%host%%` automatically getting load balanced to the respective endpoints

Reviewer: ```suggestion
- An [`http_check`][9]-based Cluster Check is dispatched to a single Agent in the cluster. This `http_check` uses the IP of the service as `%%host%%`, automatically getting load balanced to the respective endpoints.
```

> - The checks will be dispatched out with the tags `env:prod`, `service:my-nginx`, and `version:1.19.0` corresponding to the Unified Service Tagging labels

Reviewer: ```suggestion
- The checks are dispatched with the tags `env:prod`, `service:my-nginx`, and `version:1.19.0`, corresponding to Unified Service Tagging labels.
```
